File: src/cmds/generate/quote.ts
import type { WASocket } from "baileys";
import type { MessageWrapper } from "../../types/message.js";
import type { Command, CommandContext } from "../../types/cmd.js";
import { modifyExif } from "../../utils/index.js";

// Quote API request interface
interface QuoteRequest {
  type: "quote";
  format: "webp";
  isWaSticker: true;
  backgroundColor: string;
  width: number;
  height: number;
  scale: number;
  messages: Array<{
    entities: any[];
    media?: {
      base64: string;
    };
    avatar: boolean;
    from: {
      id: string | undefined;
      name: string;
      photo: {
        url?: string;
        createImageLatters?: boolean;
      };
    };
    text: string;
    replyMessage: {
      chatId?: string;
      name?: string;
      text?: string;
    };
  }>;
}

// Quote API response interface
interface QuoteResponse {
  ok: boolean;
  result: {
    image: string; // base64 string
  };
}

const command: Command = {
  name: "quote",
  description: "Generate a quote sticker from text or media",
  aliases: ["q", "qc"],
  async execute(sock, m, { args, usedPrefix, command }) {
    let media: Buffer | undefined;

    // Check if current message has media
    if (["stickerMessage", "imageMessage"].includes(m.mtype)) {
      media = await m.download() as Buffer;
    } 
    // Check if quoted message has media and no args provided
    else if (
      m.quoted &&
      ["stickerMessage", "imageMessage"].includes(m.quoted.mtype) &&
      args.length === 0
    ) {
      media = await m.quoted.download() as Buffer;
    }

    // Validate input
    const hasText = (m.quoted?.text || args.join(" ")).length > 0;
    if (!hasText && !media) {
      return m.reply(
        `Send command ${usedPrefix}${command} with text or reply to a message`
      );
    }

    // Get profile picture
    let profilePicture: string | null = null;
    const targetSender = args.length > 0 ? m.sender : m.quoted?.sender || m.sender;
    
    try {
      const ppUrl = await sock.profilePictureUrl(targetSender, "image");
      profilePicture = ppUrl || null;
    } catch (error) {
      // Profile picture not available, will use createImageLatters
    }

    // Prepare text and sender info
    const text = args.length > 0 ? args.join(" ") : m.quoted?.text || "";
    const senderId = (targetSender || m.sender).split("@")[0];
    const senderName = args.length > 0 
      ? m.pushName || `+${senderId}`
      : m.quoted?.sender
        ? `${sock.contactManager.getUser(m.quoted.sender).notify || '+' + m.quoted.senderNumber.split("@")[0]}`
        : m.pushName || `+${senderId}`;

    // Determine optimal size based on content
    const size = media ? 720 : text.length < 170 ? 520 : 1024;

    // Build reply message if applicable
    const replyMessage: { chatId?: string; name?: string; text?: string } = {};
    if (args.length > 0 && m.quoted?.text && m.quoted.text.length > 0 && m.quoted.sender) {
      const quotedSenderId = m.quoted.sender.split("@")[0];
      if (quotedSenderId) {
        // Use contactManager to retrieve pushName from quoted message
        const quotedUser = sock.contactManager.getUser(m.quoted.sender);
        const quotedName = quotedUser.notify || quotedUser.name || `+${quotedSenderId}`;
        
        replyMessage.chatId = quotedSenderId;
        replyMessage.name = quotedName;
        replyMessage.text = m.quoted.text;
      }
    }

    // Build quote request payload
    const payload: QuoteRequest = {
      type: "quote",
      format: "webp",
      isWaSticker: true,
      backgroundColor: "#1b1429",
      width: size,
      height: size,
      scale: 2,
      messages: [
        {
          entities: [],
          ...(media ? { media: { base64: media.toString("base64") } } : {}),
          avatar: true,
          from: {
            id: senderId,
            name: senderName,
            photo: profilePicture
              ? { url: profilePicture }
              : { createImageLatters: true },
          },
          text,
          replyMessage,
        },
      ],
    };

    try {
      // Call quote API using fetch
      const quoteUrl = process.env.QUOTE_URL! //|| "https://bot.lyo.su/quote/generate";
      const response = await fetch(quoteUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        throw new Error(`API request failed with status ${response.status}`);
      }

      const data: QuoteResponse = await response.json();

      if (!data.ok || !data.result?.image) {
        throw new Error("Invalid response from quote API");
      }

      // Convert base64 to buffer and add EXIF
      const imageBuffer = Buffer.from(data.result.image, "base64");
      const sticker = await modifyExif(imageBuffer);

      // Send sticker
      return sock.sendMessage(
        m.from,
        { sticker },
        { quoted: m }
      );
    } catch (error) {
      console.error("Quote generation error:", error);
      return m.reply(
        `Failed to generate quote: ${error instanceof Error ? error.message : "Unknown error"}`
      );
    }
  },
};

export default command;

File: cmds/media/quote.go
package media

import (
	"bytes"
	stdctx "context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"sync"

	hc "neo/context"
	"neo/core"
	"neo/helper"

	"go.mau.fi/whatsmeow"
	waTypes "go.mau.fi/whatsmeow/types"
)

func init() { core.Default.Register(Quote) }

// quoteAPIRequest represents the POST body for the quote-api /generate endpoint.
type quoteAPIRequest struct {
	Type            string         `json:"type"`
	Format          string         `json:"format"`
	BackgroundColor string         `json:"backgroundColor"`
	Width           int            `json:"width"`
	Height          int            `json:"height"`
	Scale           int            `json:"scale"`
	Messages        []quoteMessage `json:"messages"`
}

type quoteMessage struct {
	Entities     []any              `json:"entities"`
	Avatar       bool               `json:"avatar"`
	From         quoteFrom          `json:"from"`
	Text         string             `json:"text,omitempty"`
	Media        *quoteMedia        `json:"media,omitempty"`
	ReplyMessage *quoteReplyMessage `json:"replyMessage,omitempty"`
}

type quoteFrom struct {
	ID    string     `json:"id"`
	Name  string     `json:"name"`
	Photo quotePhoto `json:"photo"`
}

type quotePhoto struct {
	URL                string `json:"url,omitempty"`
	CreateImageLatters bool   `json:"createImageLatters,omitempty"`
}

type quoteMedia struct {
	URL string `json:"url,omitempty"`
}

type quoteReplyMessage struct {
	ChatID string `json:"chatId,omitempty"`
	Name   string `json:"name,omitempty"`
	Text   string `json:"text,omitempty"`
}

type quoteAPIResponse struct {
	Result struct {
		Image string `json:"image"`
	} `json:"result"`
	Image string `json:"image"`
}

var Quote = &core.Command{
	Name:        "quote",
	Aliases:     []string{"q", "qc"},
	Category:    "media",
	Description: "Generate a quote sticker from text or replied message",
	Run: func(ctx *hc.Ctx) {
		quoteURL := os.Getenv("QUOTE_URL")
		if quoteURL == "" {
			ctx.Reply("QUOTE_URL is not configured")
			return
		}

		args := ctx.Arguments()
		msg := ctx.Message()
		quoted := hc.GetQuotedMessage(msg)
		quotedCtxInfo := hc.GetContextInfo(msg)

		// Parse --img flag from args (accounting for smart dashes from iOS/macOS)
		sendAsImage := false
		filteredArgs := make([]string, 0, len(args))
		for _, a := range args {
			// Check for normal dashes, en dash, or em dash
			if a == "--img" || a == "-img" || a == "—img" || a == "–img" {
				sendAsImage = true
			} else {
				filteredArgs = append(filteredArgs, a)
			}
		}

		// Text args = everything except --img
		textArgs := strings.Join(filteredArgs, " ")

		// Determine text: from args or from quoted message
		text := textArgs
		if text == "" && quoted != nil {
			text = hc.GetQuotedText(quoted)
		}

		if text == "" {
			ctx.Reply(fmt.Sprintf("Send command %s%s <text> or reply to a message", ctx.Prefix(), ctx.Cmd()))
			return
		}

		// hasTextArgs = user actually typed content text (not just --img)
		hasTextArgs := textArgs != ""

		// Determine the "from" user for the quote bubble
		senderJID := ctx.SenderJID()
		senderName := ctx.Event().Info.PushName

		// Launch concurrent data fetching
		var wg sync.WaitGroup
		var mu sync.Mutex

		var quotedSenderJID waTypes.JID
		var quotedSenderName string
		hasQuotedSender := false

		// 1. Fetch quoted sender name concurrently
		wg.Add(1)
		go func() {
			defer wg.Done()
			if quoted != nil && quotedCtxInfo != nil && quotedCtxInfo.GetParticipant() != "" {
				participant := quotedCtxInfo.GetParticipant()
				jid, ok := hc.ParseJID(participant)
				if ok {
					qJID := jid.ToNonAD()
					qName := hc.GetContactName(ctx.Client(), qJID)

					mu.Lock()
					quotedSenderJID = qJID
					quotedSenderName = qName
					hasQuotedSender = true
					mu.Unlock()
				}
			}
		}()

		// 2. Resolve main sender name based on quoted status early
		if senderName == "" {
			senderName = "+" + senderJID.User
		}

		wg.Wait() // we need hasQuotedSender for the next step

		if !hasTextArgs && hasQuotedSender {
			senderJID = quotedSenderJID
			senderName = quotedSenderName
		}

		// 3. Fetch profile picture concurrently
		photoObj := quotePhoto{CreateImageLatters: true}
		wg.Add(1)
		go func(targetJID waTypes.JID) {
			defer wg.Done()
			picInfo, err := ctx.Client().GetProfilePictureInfo(
				stdctx.Background(),
				targetJID,
				&whatsmeow.GetProfilePictureParams{},
			)
			if err == nil && picInfo != nil && picInfo.URL != "" {
				mu.Lock()
				photoObj.URL = picInfo.URL
				photoObj.CreateImageLatters = false
				mu.Unlock()
			}
		}(senderJID)

		// Wait for profile picture
		wg.Wait()

		// Build replyMessage if user typed text AND is replying to a message
		var replyMsg *quoteReplyMessage
		if hasTextArgs && quoted != nil && hasQuotedSender {
			quotedText := hc.GetQuotedText(quoted)
			if quotedText != "" {
				replyMsg = &quoteReplyMessage{
					ChatID: quotedSenderJID.User,
					Name:   quotedSenderName,
					Text:   quotedText,
				}
			}
		}

		// Determine image dimensions based on text length
		size := 520
		if len(text) >= 170 {
			size = 1024
		}

		// Build request
		payload := quoteAPIRequest{
			Type:            "quote",
			Format:          "webp",
			BackgroundColor: "#1b1429",
			Width:           size,
			Height:          size,
			Scale:           2,
			Messages: []quoteMessage{
				{
					Entities: []any{},
					Avatar:   true,
					From: quoteFrom{
						ID:    senderJID.User,
						Name:  senderName,
						Photo: photoObj,
					},
					Text:         text,
					ReplyMessage: replyMsg,
				},
			},
		}

		body, err := json.Marshal(payload)
		if err != nil {
			ctx.Reply(fmt.Sprintf("Failed to build request: %v", err))
			return
		}

		resp, err := http.Post(quoteURL, "application/json", bytes.NewReader(body))
		if err != nil {
			ctx.Reply(fmt.Sprintf("Failed to call quote API: %v", err))
			return
		}
		defer resp.Body.Close()

		respBody, err := io.ReadAll(resp.Body)
		if err != nil {
			ctx.Reply(fmt.Sprintf("Failed to read response: %v", err))
			return
		}

		if resp.StatusCode != http.StatusOK {
			ctx.Reply(fmt.Sprintf("Quote API error (%d): %s", resp.StatusCode, string(respBody)))
			return
		}

		var apiResp quoteAPIResponse
		if err := json.Unmarshal(respBody, &apiResp); err != nil {
			ctx.Reply(fmt.Sprintf("Failed to parse response: %v", err))
			return
		}

		// The API may return image in result.image or directly in image
		imageB64 := apiResp.Result.Image
		if imageB64 == "" {
			imageB64 = apiResp.Image
		}
		if imageB64 == "" {
			ctx.Reply("Quote API returned empty image")
			return
		}

		imageData, err := base64.StdEncoding.DecodeString(imageB64)
		if err != nil {
			ctx.Reply(fmt.Sprintf("Failed to decode image: %v", err))
			return
		}

		if sendAsImage {
			ctx.SendImage(imageData, "")
			return
		}

		// Default: send as sticker with EXIF
		stickerData, err := helper.InjectExif(imageData, "Quote", "Bot")
		if err != nil {
			// Fallback: send raw webp as sticker
			ctx.SendSticker(imageData)
			return
		}
		ctx.SendSticker(stickerData)
	},
}
